pkg/operation/controlplanesecrets/secrets_deployer.go

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplanesecrets
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	"github.com/gardener/gardener/pkg/utils/infodata"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SecretsDeployer defines methods which can be used to deploy Secrets. The configuration from these secrets is taken from the passed in
+// GardenerResourceDataList. The deployed secrets are saved in the DeployedSecrets.
+type SecretsDeployer struct {
+	*SecretsManager
+	APIServerBasicAuth *secrets.BasicAuth
+	StaticToken        *secrets.StaticToken
+
+	ExistingSecrets map[string]*corev1.Secret
+	DeployedSecrets map[string]*corev1.Secret
+
+	mutex *sync.RWMutex
+}
+
+// NewSecretsDeployer creates a new SecretsDeployer.
+func NewSecretsDeployer(
+	gardenerResourceDataList gardencorev1alpha1helper.GardenerResourceDataList,
+	apiServerBasicAuthConfig secrets.ConfigInterface,
+	staticTokenConfig secrets.ConfigInterface,
+	certificateAuthorityConfigs map[string]*secrets.CertificateSecretConfig,
+	secretConfigGenerator SecretConfigGeneratorFunc,
+	existingSecrets map[string]*corev1.Secret,
+	deployedSecrets map[string]*corev1.Secret,
+	mutex *sync.RWMutex,
+) *SecretsDeployer {
+	s := &SecretsDeployer{SecretsManager: &SecretsManager{}}
+
+	s.CertificateAuthorities = map[string]*secrets.Certificate{}
+	s.GardenerResourceDataList = gardenerResourceDataList
+	s.APIServerBasicAuthConfig = apiServerBasicAuthConfig
+	s.StaticTokenConfig = staticTokenConfig
+	s.CertificateAuthorityConfigs = certificateAuthorityConfigs
+	s.SecretConfigGenerator = secretConfigGenerator
+	s.ExistingSecrets = existingSecrets
+	s.DeployedSecrets = deployedSecrets
+	s.mutex = mutex
+
+	return s
+}
+
+// LoadBasicAuth loads the APIServerBasicAuth secret from the corresponding InfoData
+func (s *SecretsDeployer) LoadBasicAuth() error {
```

Comment:
Hm, conceptually, I find the `Load*` functions in the `SecretsDeployer` somewhat confusing when there already is a `SecretsLoader`. Can this be improved?

Reply:
I also personally find the division of responsibilities between `SecretsManager`, `SecretsLoader`, `SecretsGenerator`, and `SecretsDeployer` a bit confusing. I think this is the root cause for some of the other comments and might cause issues in the future as well. I think here is no very good way to do such a fine-granular division in this case. I would prefer combining all these 4 types into a single one (e.g. `SecretsManager`) having all `Load`, `Generate`, and `Deploy` methods. It would turn out a bit biggish, but I see this as the "lesser evil" in this case